uniswap_edpr/src/logics/shared_algorithm_lens/src/lib.rs

```diff
@@ -0,0 +1,86 @@
+use shared_algorithm_lens_accessors :: * ;
+
+# [derive (Clone , Debug , Default , candid :: CandidType , serde :: Deserialize , serde :: Serialize)]
+pub struct LensValue {
+  pub address: String,
+  pub current_price: f32,
+  pub range_top: f32,
+  pub range_bottom: f32,
+  pub edpr: f32
+}
+
+pub async fn calculate (targets : Vec < String >) -> LensValue {
+  let pool_fees_result = get_get_last_snapshot_value_in_pool_fees (targets . get (0usize) . unwrap () . clone ()) . await ;
+  let tc28x6_result = get_get_last_snapshot_value_in_tcumul_28x6hr (targets . get (1usize) . unwrap () . clone ()) . await ;
+  let v3pool_result = get_get_last_snapshot_value_in_v3pool (targets . get (2usize) . unwrap () . clone ()) . await ;
+  let eth_usdc_price_result = get_get_last_snapshot_value_in_eth_usdc_price (targets . get (3usize) . unwrap () . clone ()) . await ;
+  
+  let fees_24h_usd = pool_fees_result.clone().unwrap().result[0].fees_24h_usd;
+  let t0_decimals = pool_fees_result.clone().unwrap().result[0].pool_summary_level_1.t0_decimals;
+  let t1_decimals =  pool_fees_result.clone().unwrap().result[0].pool_summary_level_1.t1_decimals;
+
+  let eth_price_usd  = eth_usdc_price_result.clone().unwrap().result.t1_price_usd;
+
+  let address = v3pool_result.clone().unwrap().result.address;
+  let current_tick_liquidity_str = v3pool_result.clone().unwrap().result.liquidity;
+  let current_tick_liquidity = i128::from_str_radix(current_tick_liquidity_str.trim_start_matches("0x"), 16).unwrap();
+  let sqrt_ratio_x96_str = v3pool_result.clone().unwrap().result.sqrt_ratio_x96;
+  let sqrt_ratio_x96 = u128::from_str_radix(sqrt_ratio_x96_str.trim_start_matches("0x"), 16).unwrap();
+  let tick_current = v3pool_result.clone().unwrap().result.tick_current;
+  let tick_spacing = v3pool_result.clone().unwrap().result.tick_spacing;
+  let token0 = v3pool_result.clone().unwrap().result.token0;
+
+  let tick_cumul_28x6h = tc28x6_result.unwrap().0;
+
+  let current_sqrt_price = sqrt_ratio_x96 as f32 / f32::powf(2.0,96.0);
+  let mut current_price = current_sqrt_price * current_sqrt_price;
+
+  let mut compressed: i32 = tick_current / tick_spacing;
+  if tick_current <0 && tick_current % tick_spacing != 0 {
+    compressed -= 1;
+  }
+  let floor = compressed * tick_spacing;
+  let mut min_tick = floor;
+  let mut max_tick = floor + tick_spacing;
+  for i in (1..29).rev() {
+    let new_tick = ((tick_cumul_28x6h[i] - tick_cumul_28x6h[i-1]) / 21600) as i32;
```

Comment:
I think this calculation would allow me to calculate the current tick at that time, but is it not necessary to calculate the minimum and maximum ticks for the ticks in the process that follows?

Reply:
Ideally yes, not so much for each tick value (for each 6-hour interval) but perhaps rounding the final min_tick and max_tick to the edges of the each usable tick, as shown here. This would be a bit better because liquidity needs to be provided in accordance with these boundaries. The difference in the final calculation would be very slight, however.

![image](https://github.com/horizonx-tech/chainsight-showcase/assets/37236322/658dfe3f-7908-460f-a88b-47c88d75e7ed)

Reply:
https://github.com/horizonx-tech/chainsight-showcase/pull/34#discussion_r1451373316
If my comment here is correct, I would appreciate it if you could address this one as well.
Calculating min_tick and max_tick from tick_current does not seem to be that hard amount of code.

Reply:
See my response to your comment above.

You're right - it's a relatively easy change. I tested this change at https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ab87f619a9f49dc99c15473db1ac1111. It gives a similar result, as expected. On further consideration I decided to leave it as is in the repo. My reasoning is that depending on how close the upper and lower tick values are to the upper or lower edge of their respective tick brackets, the user might decide to go either way in deciding how to choose their liquidity price range. There are other considerations to this as well (e.g. what you reckon the price will do over the coming days), so this is whether the user needs to take the information at hand and make their own decision. With this in mind, leaving it as is gives a technically more accurate and probably more helpful result.

Reply:
I understand that opinion. The problem I had was that the initial min_tick and max_tick were calculated differently than the subsequent updated values. (meaning different concepts).
If I respect your opinion, what is the advantage of calculating min_tick and max_tick from the initial current_tick?

Reply:
This literature shows that

>Lower bound of a price range pa sqrtRatioAX96 Code tracks √pa
>Upper bound of a price range pb sqrtRatioBX96 Code tracks √pb
https://atiselsts.github.io/pdfs/uniswap-v3-liquidity-math.pdf

Your opinion may have some merit in range_top and range_bottom, which are shown to the user. But for EDPR, I felt it would be better to calculate min, max for each tick. EDPR is the most important thing in this Algortithm Lens.

Reply:
In part, it's a holdover from when I was trying to add up liquidity tick by tick. In practice, it's also the narrowest price range allowed for any liquidity position so I think it's worth keeping as is for this reason. Furthermore, it also ensures that a division by zero error can't occur.

Reply:
Sorry, I didn't really understand.
I think the criteria for EDPR calculation would not match unless you decide to calculate min, max on all ticks or use current tick for all.

Reply:
https://github.com/horizonx-tech/chainsight-showcase/pull/34#issuecomment-1890476204

Reply:
https://github.com/horizonx-tech/chainsight-showcase/pull/34#issuecomment-1890478351

Reply:
>Yes, if tick_current, min_tick and max_tick are all the same then this would lead to division by zero in the amount0 and amount1 calculations. 

This can happen as long as min_tick and max_tick can be updated. Is it best to avoid this by only taking rick_spacing into account during initialization?

Reply:
>Instead we're just calculating it for the current tick bracket

This is my understanding; I only care about how min_tick and max_tick are calculated.
For the first (i.e., most recent) tick, min_tick and max_tick are the lower and upper bounds of the tick width to which current_tick belongs, taking into account tick_spacing.
For past ticks, the current_tick is directly assigned to min_tick or max_tick. I don't understand why this should not be a spacing-weighted upper or lower limit.

Reply:
>Furthermore, it also ensures that a division by zero error can't occur.
>Yes, if tick_current, min_tick and max_tick are all the same then this would lead to division by zero in the amount0 and amount1 calculations. This would hypothetically occur if there had been no trading in some pair for a full week or longer.

I thought about it a bit, but if there are no transactions for more than a week, fee = 0 and EDPR = 0%, right? If an event like the one you mentioned occurs, shouldn't it be handled explicitly so that all ticks have the same value and rdr (edpr) is 0?

Reply:
If we only use current_tick at all times,

```rust
// After loop of all ticks
if tick_current == min_tick && tick_current == max_tick {
  let price_from_tick = f32::powf(1.0001, tick_current as f32);
  return LensValue {
    address: address.to_string(),
    current_price,
    range_top: price_from_tick,
    range_bottom: price_from_tick,
    edpr: 0,
  };
}
```

I think we can avoid divided by zero by doing early returns without calculating tvl or edr.

Reply:
> I thought about it a bit, but if there are no transactions for more than a week, fee = 0 and EDPR = 0%, right?

Yes, that's right.

> If an event like the one you mentioned occurs, shouldn't it be handled explicitly so that all ticks have the same value and rdr (edpr) is 0?

I'm not sure exactly what you're asking there. Could you explain further, perhaps using the variable names to clarify further? The result would turn out as 0, as it should. This would occur with the code as it is now.

Reply:
I missed your earlier comment. Either approach would work. I'm happy to resolve this as it is.

Reply:
I don't understand. What do you mean by "the recognition is not aligned". Also, these links are not working properly on my device. Could you please repeat this post with the full quotes?

Reply:
I'm guessing that you want this calculated so that if there are no trades for a week then min_tick will equal max_tick and zero will be early-returned. I can do that if you insist on it.

Reply:
> For past ticks, the current_tick is directly assigned to min_tick or max_tick. I don't understand why this should not be a spacing-weighted upper or lower limit.

For increased precision. These are averages, so the figure for each would be less meaningful if it were to be distorted further. However, I think I now understand what you were getting at.

Reply:
On further thought, I can see a case for using current_tick is the starting value for min_tick and max_tick, which would mean that current_tick, min_tick and max_tick would all be equal if there had been no trading in a pair for a full week. With this approach, one concern is the risk that a dapp developer might use these values from the indexer and accidentally cause a division by zero error. Or does it simply behove the developer to notice this risk and takes steps to avoid it? Another concern is the practical conundrum whereby the minimum interval over which liquidity can be placed is tick_spacing, but then again the liquidity provider really just needs to make their own decision about this.

I'd like to keep (the final) min_tick and max_tick without space-weighting for the reason I gave in the previous post, but I'm happy to be persuaded either way as far as the starting values go.

Reply:
>I'd like to keep (the final) min_tick and max_tick without space-weighting for the reason I gave in the previous post, but I'm happy to be persuaded either way as far as the starting values go.

Yes, this is good.

>With this approach, one concern is the risk that a dapp developer might use these values from the indexer and accidentally cause a division by zero error. Or does it simply behove the developer to notice this risk and takes steps to avoid it?

The countermeasure to this is the error handling I have presented. Please let me know if I have misunderstood something.

https://github.com/horizonx-tech/chainsight-showcase/pull/34#discussion_r1452423466

Reply:
Thanks. I've put this in place along with a version of the error handling that you presented, slightly modified to account for the `if token0 = <ETH>` issue.

This is all good, and the function should now produce a valid output in every case. The concern I had was that if a developer used the indexer output for their own calculations within a contract, they could possibly produce a division by zero error, but I think in this case it would be the developer's responsibility to take care to avoid this.